pkg/erofs/erofs.go

```diff
@@ -0,0 +1,691 @@
+//
+// +marshal
+type InodeCompact struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Nlink        uint16
+	Size         uint32
+	Reserved     uint32
+	RawBlockAddr uint32
+	Ino          uint32
+	UID          uint16
+	GID          uint16
+	Reserved2    uint32
+}
+
+// InodeExtended represents 64-byte complete form of on-disk inode.
+//
+// +marshal
+type InodeExtended struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Reserved     uint16
+	Size         uint64
+	RawBlockAddr uint32
+	Ino          uint32
+	UID          uint32
+	GID          uint32
+	Mtime        uint64
+	MtimeNsec    uint32
+	Nlink        uint32
+	Reserved2    [16]uint8
+}
+
+// Dirent represents on-disk directory entry.
+//
+// The size of this struct is bigger than the marshalled size due to
+// the implicit padding at the end of the struct.
+//
+// +marshal
+type Dirent struct {
+	Nid      uint64
+	NameOff  uint16
+	FileType uint8
+	Reserved uint8 `marshal:"unaligned"`
+}
+
+// Image represents an open EROFS image.
+//
+// +stateify savable
+type Image struct {
+	src   *os.File `state:"nosave"`
+	bytes []byte   `state:"nosave"`
+	sb    *SuperBlock
+	root  *Inode
+}
+
+// OpenImage returns an Image providing access to the contents in the image file src.
+//
+// On success, the ownership of src is transferred to Image.
+func OpenImage(src *os.File) (*Image, error) {
+	i := &Image{src: src}
+	cu := cleanup.Cleanup{}
+
+	stat, err := i.src.Stat()
+	if err != nil {
+		return nil, err
+	}
+	i.bytes, err = syscall.Mmap(int(i.src.Fd()), 0, int(stat.Size()), syscall.PROT_READ, syscall.MAP_SHARED)
+	if err != nil {
+		return nil, err
+	}
+	cu.Add(func() { syscall.Munmap(i.bytes) })
+
+	i.sb, err = i.readSuperBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	i.root, err = i.newInode(FT_DIR, uint64(i.sb.RootNid))
+	if err != nil {
+		return nil, err
+	}
+	cu.Release()
+	return i, nil
+}
+
+// Close closes the image.
+func (i *Image) Close() {
+	syscall.Munmap(i.bytes)
+	i.src.Close()
+}
+
+// BlockSize returns the block size of this image.
+func (i *Image) BlockSize() uint32 {
+	return i.sb.BlockSize()
+}
+
+// Blocks returns the total blocks of this image.
+func (i *Image) Blocks() uint32 {
+	return i.sb.Blocks
+}
+
+// Root returns the root inode of this image.
+func (i *Image) Root() *Inode {
+	return i.root
+}
+
+// readSuperBlock returns the super block of this image.
+func (i *Image) readSuperBlock() (*SuperBlock, error) {
+	sb := &SuperBlock{}
+	sb.UnmarshalUnsafe(i.bytes[SUPER_BLOCK_OFFSET:])
+
+	if sb.Magic != SUPER_BLOCK_MAGIC_V1 {
+		return nil, fmt.Errorf("unknown magic: 0x%x", sb.Magic)
+	}
+
+	if err := i.verifyChecksum(*sb); err != nil {
+		return nil, err
+	}
+
+	if featureIncompat := sb.FeatureIncompat & ^uint32(FEATURE_INCOMPAT_ALL); featureIncompat != 0 {
+		return nil, fmt.Errorf("unsupported incompatible features detected: 0x%x", featureIncompat)
+	}
+
+	if sb.BlockSize() != hostarch.PageSize {
```

Comment:
by the way, currently uncompressed erofs doesn't have such limitation now (assuming this implementation doesn't have such limitation either?)

Reply:
Thanks for the heads-up. I thought erofs always requires that the block size equals to the page size. This implementation doesn’t have this limitation, but it does require that each block is page size aligned, as we want to map the data on the image file to the guest userspace directly. If erofs doesn’t have such limitation, we will just need to check `sb.BlockSize() % hostarch.PageSize == 0`.

Reply:
Formally, erofs supports blocksize <= pagesize because currently linux storage I/O stack limits that.
But for your use cases, I'm fine with "sb.BlockSize() % hostarch.PageSize == 0" (since we could then handle 64k-block images generated in 64k page host, but we can then also use it if we switch to 4k page size as well.)

Reply:
I see. Thanks for the details.